subgraphs/_reference_/src/sdk/protocols/perpfutures/pool.ts

```diff
@@ -0,0 +1,810 @@
+   */
+  addTotalValueLocked(delta: BigDecimal): void {
+    this.pool.totalValueLockedUSD = this.pool.totalValueLockedUSD.plus(delta);
+    this.save();
+
+    this.protocol.addTotalValueLocked(delta);
+  }
+
+  /**
+   * Recalculates the total value locked for this pool based on its current input token balance.
+   * This function will also update the protocol's total value locked based on the change in this pool's.
+   */
+  refreshTotalValueLocked(): void {
+    let totalValueLockedUSD = constants.BIGDECIMAL_ZERO;
+
+    for (let idx = 0; idx < this.pool.inputTokens.length; idx++) {
+      const inputTokenBalance = this.pool.inputTokenBalances[idx];
+      const inputToken = this.tokens.getOrCreateTokenFromBytes(
+        this.pool.inputTokens[idx]
+      );
+
+      const amountUSD = this.getInputTokenAmountPrice(
+        inputToken,
+        inputTokenBalance
+      );
+      totalValueLockedUSD = totalValueLockedUSD.plus(amountUSD);
+    }
+
+    this.setTotalValueLocked(totalValueLockedUSD);
+  }
+
+  /**
+   * Updates the total value locked for this pool to the given value.
+   * Will also update the protocol's total value locked based on the change in this pool's.
+   */
+  setVolume(newVolume: BigDecimal): void {
+    const delta = newVolume.minus(this.pool.cumulativeVolumeUSD);
+    this.addVolume(delta);
+    this.save();
+  }
+
+  /**
+   * Adds the given delta to the total value locked for this pool.
+   * Will also update the protocol's total value locked based on the change in this pool's.
+   *
+   * @param delta The change in total value locked for this pool.
+   */
+  addVolume(delta: BigDecimal): void {
+    this.pool.cumulativeVolumeUSD = this.pool.cumulativeVolumeUSD.plus(delta);
+    this.save();
+
+    this.protocol.addVolume(delta);
+  }
+
+  /**
+   * Adds a given USD value to the pool and protocol supplySideRevenue. It can be a positive or negative amount.
+   * Same as for the rest of setters, this is mostly to be called internally by the library.
+   * But you can use it if you need to. It will also update the protocol's snapshots.
+   * @param rev {BigDecimal} The value to add to the protocol's supplySideRevenue.
+   */
+  addSupplySideRevenueUSD(rev: BigDecimal): void {
+    this.pool.cumulativeTotalRevenueUSD =
+      this.pool.cumulativeTotalRevenueUSD.plus(rev);
+    this.pool.cumulativeSupplySideRevenueUSD =
+      this.pool.cumulativeSupplySideRevenueUSD.plus(rev);
+    this.save();
+
+    this.protocol.addSupplySideRevenueUSD(rev);
+  }
+
+  /**
+   * Adds a given USD value to the pool and protocol protocolSideRevenue. It can be a positive or negative amount.
+   * Same as for the rest of setters, this is mostly to be called internally by the library.
+   * But you can use it if you need to. It will also update the protocol's snapshots.
+   * @param rev {BigDecimal} The value to add to the protocol's protocolSideRevenue.
+   */
+  addProtocolSideRevenueUSD(rev: BigDecimal): void {
```

Comment:
The comment says it updates the snapshot, but I don't see the code for that

Reply:
Same question for most of the following functions

Reply:
Ah I see, the snapshots are literally taken by calling `takeSnapshot()`. Are there functions in the SDK that we can make those calls from so the user doesn't need to touch it?